### PR TITLE
Make `StoreOp` an enum instead of a bool

### LINKED ISSUE #4146: Depth target gets cleared before compute pass when previously used as attachement without storing values
**Description**
Occurs with the following sequence

* render pass that clears targets and stores depth values
* render pass that does not clear targets and does not store depth values
* (wgpu clears depth target internally)
* depth target gets used as texture_depth_2d in compute pass (and has now all 0 values) -> I would have expected to still have the depth values from the first render pass available here

**Repro steps**
see the attached example

**Expected vs observed behavior**
I would expect that the depth target does not get cleared when running the compute pass.

**Extra materials**
The attached zip contains a minimal example of the issue.

**Platform**
Windows 10
wgpu 0.17 and master on Vulkan backend
RX 570 driver version 21.50.21.11

[wgpu-depth-clear-bug.zip](https://github.com/gfx-rs/wgpu/files/12640986/wgpu-depth-clear-bug.zip)

### PR DESCRIPTION
**Pull request by Wumpf**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
* fixes #4146
   * (description is actually not indicative of the actual error but of the common misconception this avoids)

**Description**

`wgpu::Operations::store` used to be an underdocumented boolean value,
causing missunderstandings of the effect of setting it to `false`.

The API now resembles WebGPU which distinguishes `store` and `discard`,
see [WebGPU spec on GPUStoreOp](https://gpuweb.github.io/gpuweb/#enumdef-gpustoreop).

```diff
// ...
depth_ops: Some(wgpu::Operations {
    load: wgpu::LoadOp::Clear(1.0),
-   store: false,
+   store: wgpu::StoreOp::Discard,
}),
// ...
```

Opted to put this text into the changelog under `Major Changes` since it breaks like every single use ;-)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -64,6 +64,28 @@ let render_pass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
 
 By @Valaphee in [#3402](https://github.com/gfx-rs/wgpu/pull/3402)
 
+
+#### Render pass store operation is now an enum
+
+`wgpu::Operations::store` used to be an underdocumented boolean value,
+causing missunderstandings of the effect of setting it to `false`.
```
Comment (cwfitzgerald) [inline CHANGELOG.md]:
```suggestion
causing misunderstandings of the effect of setting it to `false`.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -64,6 +64,28 @@ let render_pass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
...
+
+The API now resembles WebGPU which distinguishes `store` and `discard`,
+see [WebGPU spec on GPUStoreOp](https://gpuweb.github.io/gpuweb/#enumdef-gpustoreop).
```
Comment (cwfitzgerald) [inline CHANGELOG.md]:
```suggestion
The API now more closely resembles WebGPU which distinguishes between `store` and `discard`,
see [WebGPU spec on GPUStoreOp](https://gpuweb.github.io/gpuweb/#enumdef-gpustoreop).
```

File: wgpu/src/lib.rs [hunk 3]
```diff
@@ -1006,16 +1006,22 @@ static_assertions::assert_impl_all!(BufferBinding: Send, Sync);
...
 pub enum LoadOp<V> {
-    /// Clear with a specified value.
+    /// Loads the specified value for this attachment into the render pass.
+    ///
+    /// On some GPU hardware (primarily mobile), "clear" is significantly cheaper
+    /// because it avoids loading data from main memory into tile-local memory.
+    /// On other GPU hardware, there isn’t a significant difference.
+    /// As a result, it is recommended to use "clear" rather than "load" in cases
+    /// where the initial value doesn’t matter
+    /// (e.g. the render target will be cleared using a skybox).
```
Comment (cwfitzgerald) [inline wgpu/src/lib.rs]:
```suggestion
    /// Loads the specified value for this attachment into the render pass.
    ///
    /// On some GPU hardware (primarily mobile), "clear" is significantly cheaper
    /// because it avoids loading data from main memory into tile-local memory.
    ///
    /// On other GPU hardware, there isn’t a significant difference.
    ///
    /// As a result, it is recommended to use "clear" rather than "load" in cases
    /// where the initial value doesn’t matter
    /// (e.g. the render target will be cleared using a skybox).
```

Just spacing to make paragraph breaks

File: wgpu/src/lib.rs [hunk 4]
```diff
@@ -1025,6 +1031,25 @@ impl<V: Default> Default for LoadOp<V> {
...
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Default)]
+#[cfg_attr(feature = "trace", derive(serde::Serialize))]
+#[cfg_attr(feature = "replay", derive(serde::Deserialize))]
+pub enum StoreOp {
+    /// Stores the resulting value of the render pass for this attachment.
+    #[default]
+    Store,
+    /// Discards the resulting value of the render pass for this attachment.
+    ///
+    /// The attachment/affected texture aspect(s) will be treated as uninitialized afterwards.
```
Comment (cwfitzgerald) [inline wgpu/src/lib.rs]:
```suggestion
    /// The attachment's affected texture aspect(s) will be treated as uninitialized afterwards.
```

This.. might be clearer? Not a lot of people understand texture aspects, so I'm hesitant to write about it without something explaining what it is. Maybe a link back to the texture aspect type?

File: wgpu/src/lib.rs [hunk 5]
```diff
@@ -1025,6 +1031,25 @@ impl<V: Default> Default for LoadOp<V> {
...
+    ///
+    /// This can be significantly faster on tile-based render hardware.
+    /// Prefer this if the attachment is not read by subsequent passes.
```
Comment (cwfitzgerald) [inline wgpu/src/lib.rs]:
```suggestion
    /// This can be significantly faster on tile-based render hardware.
    ///
    /// Prefer this if the attachment is not read by subsequent passes.
```

File: wgpu/src/lib.rs [hunk 6]
```diff
@@ -1036,14 +1061,15 @@ pub struct Operations<V> {
     /// How data should be read through this attachment.
     pub load: LoadOp<V>,
     /// Whether data will be written to through this attachment.
```
Comment (cwfitzgerald) [inline wgpu/src/lib.rs:1068]: I think we also need to change this comment/the other one to mention the fact that the resolve texture is _always_ stored to, no matter what this is set to.

File: wgpu/src/backend/direct.rs [hunk 7]
```diff
@@ -392,6 +392,13 @@ fn map_texture_tagged_copy_view(
     }
 }
 
+fn map_store_op(op: StoreOp) -> wgc::command::StoreOp {
+    match op {
+        StoreOp::Store => wgc::command::StoreOp::Store,
+        StoreOp::Discard => wgc::command::StoreOp::Discard,
+    }
+}
```
Comment (cwfitzgerald) [inline wgpu/src/backend/direct.rs:400]: We totally should lift this into wgt and use the same type everywhere

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Comments, like the api change!

File: wgpu/src/lib.rs (hunk 4 above)
Comment (Wumpf) [reply wgpu/src/lib.rs]: we could also drop it from here, it's a super strange implementation detail about how different aspects of a texture can be in an uninitialized state (I hate it ;))

File: wgpu/src/backend/direct.rs (hunk 7 above)
Comment (Wumpf) [reply wgpu/src/backend/direct.rs:400]: I was about to, but then figured since the LoadOp can't go into wgt (it has the clear value with it for improved Rust ergonomics) I didn't want to put them into different places so it's easier to find stuff 🤔

**Comment by Wumpf** [comment]
Thanks for all the comment improvements and keeping the bar for documentation high!

File: wgpu/src/lib.rs (hunk 6 above)
Comment (Wumpf) [reply wgpu/src/lib.rs:1068]: good idea

File: wgpu/src/backend/direct.rs (hunk 7 above)
Comment (cwfitzgerald) [reply wgpu/src/backend/direct.rs:400]: Fair

File: wgpu/src/lib.rs (hunk 6 above)
Comment (cwfitzgerald) [reply wgpu/src/lib.rs:1068]: This doesnt' seem to have been addressed.

File: wgpu/src/lib.rs (hunk 4 above)
Comment (cwfitzgerald) [reply wgpu/src/lib.rs]: It's not really an implementation detail, as it results in a visible clear happening - maybe we should document that behavior.

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Did you miss a push, a couple resolved things are still unaddressed

**Comment by Wumpf** [comment]
> Did you miss a push

eh, yeah I managed to push to the wrong remote 🤦

File: wgpu/src/lib.rs (hunk 4 above)
Comment (Wumpf) [reply wgpu/src/lib.rs]: Hm maybe just best to talk about depth/stencil aspects inline here, shouldn't be too hard, let me try.

**Review by teoxoy — APPROVED** [review]
LGTM!

File: wgpu/src/lib.rs [hunk 8]
```diff
@@ -1036,14 +1066,18 @@ pub struct Operations<V> {
...
-    pub store: bool,
+    ///
+    /// Note that resolve textures (if specified) are always written to,
+    /// regardless of this setting.
+    pub store: StoreOp,
 }
 
 impl<V: Default> Default for Operations<V> {
+    #[inline]
```
Comment (cwfitzgerald) [inline wgpu/src/lib.rs:1076]: Why was this added?

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]

File: wgpu/src/lib.rs (hunk 8 above)
Comment (Wumpf) [reply wgpu/src/lib.rs:1076]: counter question, why was it missing - I added this in passing since trivial default like this should imho be inlinable and Rust's compilation model doesn't allow for inlining accros crates unless either this attribute is present or link time optimizations are at play.
Unrelated to this PR ofc, but figured can't hurt. You object?

Comment (cwfitzgerald) [reply wgpu/src/lib.rs:1076]: I'm honestly not sure of the full costs/benefits of this - I know it can benefit when you have LTO off (but also why do you have LTO off if you care about performance) but I think there is costs to just slapping it on all over the place, even for trivial stuff.

Comment (cwfitzgerald) [reply wgpu/src/lib.rs:1076]: Not a discussion for this PR - let's merge.

**Review by cwfitzgerald — APPROVED** [review]
